src/sinks/mqtt/sink.rs

```diff
@@ -0,0 +1,215 @@
+            connector,
+            finalizers_queue,
+            quality_of_service: config.quality_of_service,
+            events_sent: register!(EventsSent::from(Output(None))),
+            bytes_sent: register!(BytesSent::from(Protocol("mqtt".into()))),
+        })
+    }
+
+    /// outgoing events main loop
+    async fn handle_events<I>(
+        &mut self,
+        input: &mut I,
+        client: &mut AsyncClient,
+        connection: &mut EventLoop,
+    ) -> Result<(), ()>
+    where
+        I: Stream<Item = Event> + Unpin,
+    {
+        loop {
+            tokio::select! {
+                // handle connection errors
+                msg = connection.poll() => {
+                    match msg {
+                        Ok(rumqttc::Event::Outgoing(rumqttc::Outgoing::PubRel(_))) => {
+                            // publish has been acknowledged by the MQTT server
+                            if let Some(finalizers) = self.finalizers_queue.pop_front() {
+                                finalizers.update_status(EventStatus::Delivered);
+                            }
+                        }
+                        Ok(_) => {}
+                        Err(error) => {
+                            emit!(MqttConnectionError { error });
+                            return Err(());
+                        }
+                    }
+                },
+
+                // handle outgoing events
+                event = input.next() => {
+                    let mut event = if let Some(event) = event {
+                        event
+                    } else {
+                        break;
+                    };
+
+                    let finalizers = event.take_finalizers();
+
+                    let topic = match self.connector.topic.render_string(&event) {
+                        Ok(topic) => topic,
+                        Err(error) => {
+                            emit!(TemplateRenderingError {
+                                error,
+                                field: Some("topic"),
+                                drop_event: true,
+                            });
+                            finalizers.update_status(EventStatus::Errored);
+                            continue;
+                        }
+                    };
+
+                    self.transformer.transform(&mut event);
+
+                    let event_byte_size = event.size_of();
+
+                    let mut bytes = BytesMut::new();
+                    let message = match self.encoder.encode(event, &mut bytes) {
+                        Ok(()) => {
+                            bytes.to_vec()
+                        }
+                        Err(_) => {
+                            finalizers.update_status(EventStatus::Errored);
+                            continue;
+                        }
+                    };
+                    let message_len = message.len();
+
+                    let qos = match self.quality_of_service {
```

Comment:
Very minor point.. We could have a `impl From<MqttQos> for Qos` for this.

Reply:
Fixed.